### PR TITLE
Porting tests to new utest-greentea framework

### PR DESCRIPTION
**Pull request by mazimkhan**
Following tests are ported to new utest-greentea framework. Also necessary changes have be done in the test code for tests to work and for clean tear down.
```
sockets-test-echo-tcp-client
sockets-test-echo-udp-client
```
For details on new utest - greentea framework please see description in PR https://github.com/ARMmbed/greentea/pull/78

### CODE REVIEW

**Comment by mazimkhan** [comment]
@bremoran @niklas-arm @PrzemekWirkus @0xc0170 @stegut01
Please review this PR. It is port for new utest-greentea framework. There might be some understanding gaps on socket and utest API. Guidance and documentation would be much appreciated.

File: test/echo-tcp-client/main.cpp [hunk 1]
```diff
@@ -41,6 +36,7 @@ void terminate(bool status, TCPEchoClient* client);
 
 char out_buffer[] = "Hello World\n";
 char buffer[256];
+char out_success[] = "{{success}}\n{{end}}\n";
```
Comment (PrzemekWirkus) [inline test/echo-tcp-client/main.cpp]: This string is confusing, is this just a payload used for test or actual control command?

**Comment by mazimkhan** [comment]
CI build failing for error
```
error: mbed-drivers does not meet specification ~0.12.0 required by core-util
```

File: test/echo-tcp-client/main.cpp (hunk 1 above)
Comment (mazimkhan) [inline test/echo-tcp-client/main.cpp]: Moved it outside from onRx() in original code to sit with other strings. It indicates to the host test that target test has validated a round trip (echo) data and now terminating. It is good stop host test from echoing back as target client closes the connection right after sending this string. 
But it should not be confused with serial k,v command.

**Comment by niklas-arm** [comment]
I might be mistaken, but isn't TCP and UDP client already part of socket-test?
They are implemented from scratch, but they test the same thing

- [TCP client (HTTP request)](https://github.com/ARMmbed/socket-test/blob/master/test/tcp_download_v0/main.cpp)
- [UDP client (Time Protocol)](https://github.com/ARMmbed/socket-test/blob/master/test/time_protocol_v0/main.cpp)

I did take the tests in this repo as a basis for the new tests, but with utest expressing the statefulness of these tests can be achieved much better without classes.

**Comment by 0xc0170** [comment]
@niklas-arm Yes, I asked Przemek the same question. I was told that all these public tests which are part of our modules will be updated to use the new greentea client, deprecate to use test_env. Plus, those tests you references, are not public yet. Przemek might add more details.

**Comment by mazimkhan** [comment]
@niklas-arm if tests were moved to socket-test then as part of the move they should have been deleted from this repo. So that nobody wastes time in fixing or running them. Please remove these tests from this repo if not needed here now.

Anyway other purpose of updating these tests is to gain understanding of utest. You can atleast review from the perspective of utest usage.

**Comment by niklas-arm** [comment]
> @niklas-arm if tests were moved to socket-test then as part of the move they should have been deleted from this repo. So that nobody wastes time in fixing or running them. Please remove these tests from this repo if not needed here now.

The tests weren't "moved", we just didn't make the new ones public, and we didn't update these.
Let's not remove these rewritten tests now, since they are slightly different from the HTTP request and Time Protocol test, which do not use local Servers.
So let's keep these tests, especially as an example for how to use new host test mechanism (which look great btw!).

> Anyway other purpose of updating these tests is to gain understanding of utest. You can atleast review from the perspective of utest usage.

I would have liked to have a more granular test cases, instead of one big class, similar to how socket-tests work, but if it works this way, it's fine too.

**Comment by mazimkhan** [comment]
Good to know that the effort isn't wasted. We can surely have granular tests in socket-test/test.
@bremoran if you are happy with the change can you please merge it.

File: module.json [hunk 2]
```diff
@@ -21,7 +21,9 @@
     "minar": "^1.0.0"
   },
   "testDependencies": {
-    "mbed-drivers": "~0.11.1"
+    "mbed-drivers": "~0.11.1",
```
Comment (bremoran) [inline module.json]: This dependency is a little old.

File: test/echo-tcp-client/main.cpp [hunk 3]
```diff
@@ -76,65 +77,55 @@ class TCPEchoClient {
...
-            printf("%s\r\n", buffer);
-            rc = strncmp(out_buffer, buffer, sizeof(out_buffer) - 1);
-            if (TEST_EQ(rc, 0)) {
-                _unacked += sizeof(out_success) - 1;
-                err = _stream.send(out_success, sizeof(out_success) - 1);
-                _done = true;
-                TEST_EQ(err, SOCKET_ERROR_NONE);
-                minar::Scheduler::postCallback(fpterminate_t(terminate).bind(TEST_RESULT(),this));
-            }
+            int rc = strncmp(out_buffer, buffer, sizeof(out_buffer) - 1);
```
Comment (bremoran) [inline test/echo-tcp-client/main.cpp]: This could be replaced with TEST_ASSERT_EQUAL_ARRAY

**Comment by bremoran** [comment]
A couple of minor comments. updating the dependency to ~0.12.0 will fix the CI build error.

**Comment by mazimkhan** [comment]
@bremoran review comments incorporated.

**Pull request merged**